offchain/danaswap-api/src/DanaSwap/Api.purs

```diff
@@ -9,30 +17,119 @@ import Contract.Prelude
 import Contract.Address (getWalletAddress, getWalletCollateral, scriptHashAddress)
 import Contract.Log (logDebug', logInfo')
 import Contract.Monad (Contract, liftContractM)
-import Contract.PlutusData (Datum(..), PlutusData(..), toData)
+import Contract.PlutusData (Datum(..), PlutusData(..), Redeemer(..), toData)
+import Contract.Prim.ByteArray (hexToByteArray)
 import Contract.ScriptLookups as Lookups
 import Contract.Scripts (MintingPolicy, Validator, mintingPolicyHash, validatorHash)
-import Contract.Transaction (TransactionInput)
+import Contract.Transaction (TransactionInput, TransactionOutputWithRefScript)
 import Contract.TxConstraints (DatumPresence(..), TxConstraints)
 import Contract.TxConstraints as Constraints
 import Contract.Utxos (getUtxo)
-import Contract.Value (CurrencySymbol, adaToken, mpsSymbol, scriptCurrencySymbol)
+import Contract.Value (CurrencySymbol, TokenName, adaToken, mkTokenName, mpsSymbol, scriptCurrencySymbol, symbols, valueOf)
 import Contract.Value as Value
 import Ctl.Util (buildBalanceSignAndSubmitTx, getUtxos, waitForTx)
 import DanaSwap.CborTyped (configAddressValidator, liqudityTokenMintingPolicy, poolAddressValidator, poolIdTokenMintingPolicy, simpleNft)
 import Data.BigInt as BigInt
 import Data.List (head)
-import Data.Map (keys)
+import Data.Map (Map, keys)
 import Data.Map as Map
 import Data.Set (toUnfoldable)
+import Effect.Exception (throw)
 
 type Protocol =
   { configUtxo :: TransactionInput
-  , poolVal :: Validator
+  , poolAdrVal :: Validator
   , liquidityMP :: MintingPolicy
   , poolIdMP :: MintingPolicy
   }
 
+type PoolId = TokenName
+
+getAllPools :: Protocol -> Contract () (Map TransactionInput TransactionOutputWithRefScript)
+getAllPools protocol@{ poolAdrVal } =
+  getUtxos (scriptHashAddress $ validatorHash poolAdrVal)
+    <#> Map.filter (hasNft protocol)
```

Comment:
I imagine there is some complexity here. Is it possible to paginate?

Reply:
I've not seen the word paginate before, it seems to mean to add page numbers? I'm not sure what you mean.

Reply:
Oh okay,  AFAICT no we'd need CTL support for that
(you need to serve offchain docs for this link) 
http://localhost:8080/Contract.Utxos.html#v:utxosAt
there doesn't seem to be a paginate verion of this function that would be core to this.
I guess we could do take 15 or something and count on laziness for the filtering to be done on demand?

Reply:
This is definitely a scaling issue, which we'll either have to solve through indexing via the pools aggregation in DanaSwap Stats, or ensure pool initialization fees are high enough to decentivize the creation of dust pools. The idealist in me favors the former, putting the onus on the dApp to define it's query model. Blockchains after all are not query optimized, nor should be.

cc @CSVdB

Reply:
Could you include a TODO here to remove this function and replace it with a call to the Stats endpoint once it's supported?